### PR TITLE
Refuse to replace already replaced values

### PR DESCRIPTION
**Pull request by mantoni**
Fix for issue #1779

When values where replaced multiple times, `restore` didn't restore the original value. This change causes `sandbox.replace` to throw if the given object / property combination was already used.

#### How to verify - mandatory
1. Check out this branch
2. `npm install`
3. `npm test`

or

```js
sinon.replace(console, "log", sinon.fake());
sinon.replace(console, "log", sinon.fake()); // now throws
```

## Versioning

While this fixes an issue where `sandbox.restore()` fails to restore values that where replaced multiple times, this might also cause existing tests to fail. However, I would argue that we're fixing a broken behavior in the new `sandbox.replace` API to bring it in line with `sandbox.stub` which also throws in this case. Therefore I'd vote for semver patch.

#### Checklist for author

- [x] `npm run lint` passes
- [x] References to standard library functions are [cached](https://github.com/sinonjs/sinon/pull/1523).

### CODE REVIEW

File: lib/sinon/stub.js [hunk 1]
```diff
@@ -104,10 +104,10 @@ var proto = {
             return getCurrentBehavior(fnStub).invoke(this, arguments);
         };
 
-        functionStub.id = "stub#" + uuid++;
         var orig = functionStub;
         functionStub = spy.create(functionStub, stubLength);
         functionStub.func = orig;
+        functionStub.id = "stub#" + uuid++;
```
Comment (mantoni) [inline lib/sinon/stub.js]: Uh, sorry. Commited this by accident. I noticed that the generated `stub#${uuid}` id property is replaced by `spy.create` with `spy#{uuid}`. This change fixes it, but it's not really part of this PR. Shall I revert or just leave it?

File: lib/sinon/sandbox.js [hunk 2]
```diff
@@ -161,9 +161,21 @@ function Sandbox() {
...
             Object.defineProperty(object, property, descriptor);
-        };
+        }
+        restorer.object = object;
+        restorer.property = property;
+        return restorer;
+    }
+
+    function verifyNotReplaced(object, property) {
+        for (var i = 0; i < fakeRestorers.length; i++) {
```
Comment (mroderick) [inline lib/sinon/sandbox.js]: Since we're targeting ES5.1, it's fine to use the ES5 Array Extras... it is much harder to create off-by-one errors, when you don't need to maintain an index 😉 

```js
function verifyNotReplaced(object, property) {
    fakeRestorers.forEach(function(fakeRestorer) {
        if (fakeRestorer.object === object && fakeRestorer.property === property) {
            throw new TypeError("Attempted to replace " + property + " which is already replaced");
        }
    });
}
```

File: lib/sinon/stub.js (hunk 1 above)
Comment (mroderick) [reply lib/sinon/stub.js]: Maybe just put it into it's own commit?

File: lib/sinon/sandbox.js (hunk 2 above)
Comment (mantoni) [reply lib/sinon/sandbox.js]: Yeah, you're right. And even `forEach` would give us an index as the second argument, if needed.

**Comment by mroderick** [comment]
This has been published as `sinon@5.0.5`
